### PR TITLE
Virt network create

### PR DESCRIPTION
**Pull request by cbosdo**
## What does this PR change?

Add virtual network creation action.

## GUI diff

Add an action and the corresponding dialog to create a virtual network.

Before:

Actions before

After:

Actions after

Dialog example

- [X] **DONE**

## Documentation
- Documentation issue was created: [Link for SUSE Manager contributors](https://github.com/SUSE/spacewalk/issues/new?template=ISSUE_TEMPLATE_DOCUMENTATION.md&labels=documentation&projects=SUSE/spacewalk/31), [Link for community contributors](https://github.com/uyuni-project/uyuni-docs/issues/new).
- (OPTIONAL) [Documentation PR](https://github.com/uyuni-project/uyuni-docs/pulls)

- [ ] **DONE**

## Test coverage
- Unit tests were added
- Cucumber tests were added

- [ ] **DONE**

## Links

Fixes #
Tracks # **add downstream PR, if any**

- [X] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

File: branding/css/theme.less [hunk 1]
```diff
@@ -890,3 +890,58 @@ code {
...
+  .column-title {
+    font-weight: bolder;
+    margin-left: -15px;
+    margin-right: -15px;
+    padding-left: 30px;
+  }
+}
+
+// Represents network interfaces
+.interface {
```
Comment (Etheryte) [inline branding/css/theme.less:905]: Here and below, I would recommend extracting these into a separate file and importing that in this file. These are fairly specific and this file is already pretty large on its own right.

File: web/html/src/components/input/Form.tsx [hunk 2]
```diff
@@ -137,12 +137,14 @@ export class Form extends React.Component<Props> {
...
-            return filtered;
-          }, []);
+          const errors = this.props.errors != null ?
+            Object.keys(this.props.errors || {}).reduce((filtered, key) => {
+              if (names.includes(key) && this.props.errors?.[key]) {
+                return filtered.concat(this.props.errors[key]);
+              }
+              return filtered;
+            }, []) :
+            undefined;
```
Comment (Etheryte) [inline web/html/src/components/input/Form.tsx:147]: It's not clear what the intent here is. What's the benefit of having `undefined` instead of an empty array `[]` annotating no errors?

File: web/html/src/components/input/FormGroup.tsx [hunk 3]
```diff
@@ -3,8 +3,16 @@ import * as React from "react";
...
 };
 
 export function FormGroup(props: Props) {
-  return <div className={`form-group${props.isError ? " has-error" : ""}`}>{props.children}</div>;
+  const className = props.className != null ? props.className : '';
+  return <div className={`form-group${props.isError ? " has-error" : ""} ${className}`}>{props.children}</div>;
+}
+
+FormGroup.defaultProps = {
+  className: undefined,
```
Comment (Etheryte) [inline web/html/src/components/input/FormGroup.tsx:17]: If the default value is `undefined` there's no need to add this in my opinion, the default value of everything that doesn't exist is `undefined`.

File: web/html/src/components/input/InputBase.tsx [hunk 4]
```diff
@@ -173,7 +177,10 @@ export class InputBase<ValueType = string> extends React.Component<InputBaseProp
     }
 
     if (!this.props.disabled && (value || this.props.required)) {
-      if (this.props.required && !value) {
+      const noValue = Array.isArray(this.props.name) ?
+        value == null || Object.values(value).filter(v => v).length === 0 :
```
Comment (Etheryte) [inline web/html/src/components/input/InputBase.tsx:181]: Using `==` is a linter warning, and should ideally be an error instead (but I need to clean up all instances first). I know it's convenient to use to check for both `undefined` and `null` in one go, but the problem is it isn't clear whether that's the intent or it's just a typo.  
I would recommend using `isNil` from lodash instead in this case.

Comment (Etheryte) [inline web/html/src/components/input/InputBase.tsx:181]: `filter(v => v)` is a dangerous way to solve this since `InputBase` is a generic. We currently don't have such a case, but it would be completely valid to use `InputBase` in a configuration where the value type is a number or boolean, in which this filter will result in a bug.  

As a fix I would recommend adding a method to the class a-la `isEmptyValue` that's something along these lines:  
```ts
isEmptyValue(input: any) {
  if (typeof input === 'string') {
    return input.trim() === '';
  }
  return isNil(input);
}
```

File: web/html/src/components/input/InputBase.tsx [hunk 5]
```diff
@@ -173,7 +177,10 @@ export class InputBase<ValueType = string> extends React.Component<InputBaseProp
...
+        !value;
```
Comment (Etheryte) [inline web/html/src/components/input/InputBase.tsx:182]: Similar to the other comments above, it would be better to be explicit with the intent here, perhaps reusing the `isEmptyValue` idea.

File: web/html/src/components/input/form-utils.ts [hunk 6]
```diff
@@ -76,6 +78,7 @@ export function unflattenModel<T>(flatModel: Record<string, T>): TreeLikeModel<T
 
   Object.keys(flatModel)
     .sort()
+    .filter(name => !stripBlankValues || ((flatModel[name] || "") !== ""))
```
Comment (Etheryte) [inline web/html/src/components/input/form-utils.ts:81]: I would recommend extracting this part into a separate function and keep the signature for `unflattenModel` as is. That allows us to keep a fairly clean separation of concerns, where one function does one thing.

Comment (Etheryte) [inline web/html/src/components/input/form-utils.ts:81]: `(flatModel[name] || "") !== ""` will not work as intended when the value is a number or a boolean.  

```ts
(0 || "") !== ""; // false
(1 || "") !== ""; // true
(false || "") !== ""; // false
(true || "") !== ""; // true
```

Similar to the suggestions in `InputBase`, I would recommend being explicit about what to do with what input instead.

File: web/html/src/manager/virtualization/nets/network-properties.js [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+
+    const toClean = ['interface-selection-type', 'ipv6-enabled', 'virtualport-params-type']
+    toClean.forEach(field => {
+      if (definition[field] != null) {
+        delete definition[field];
+      }
+    });
+
+    // Rename the ipv4def and ipv6def properties
+    ["ipv4", "ipv6"].forEach(prop => {
```
Comment (Etheryte) [inline web/html/src/manager/virtualization/nets/network-properties.js:82]: Perhaps I don't understand the intent here, why can't these fields have the intended name right away?

File: web/html/src/manager/virtualization/nets/network-properties.js [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+                    divClass="col-md-6"
+                  />)
+                }
+                <Select
+                  labelClass="col-md-3"
+                  divClass="col-md-6"
+                  label={t('Network type')}
+                  name="type"
+                  required
+                  disabled={(props.initialModel || {}).type !== undefined}
```
Comment (Etheryte) [inline web/html/src/manager/virtualization/nets/network-properties.js:168]: I would recommend either using `typeof foo !== 'undefined'` instead, this configuration is safe now, but if the condition flips in the future, simply changing `!==` to `===` in the current code will create a runtime error.

File: web/html/src/manager/virtualization/nets/network-properties.js [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+                    {
+                      (model.interfaces || []).map(nic => (
+                        <div className="col-md-6 col-md-offset-3">
+                          <Interface device={netDevices.find(dev => dev.name === nic)} />
+                        </div>
+                      ))
+                    }
+                  </>
+                )}
+                { model['interface-selection-type'] === 'pf' && FieldsData.getValue(model.type, 'pf', false) && (
```
Comment (Etheryte) [inline web/html/src/manager/virtualization/nets/network-properties.js:277]: I would recommend extracting these bigger blocks out of the template into a boolean variable with a descriptive name a-la `showPhysicalFunction`. Right now it's a bit of a mouthful to read the template since there's a lot of logic and view stuff mixed in in long goes.

File: web/html/src/manager/virtualization/nets/properties/NetworkAddress.js [hunk 10]
```diff
@@ -0,0 +1,107 @@
+// @flow
+import * as React from 'react';
+import Validation from 'components/validation';
+import { InputBase } from 'components/input/InputBase';
+import { FormContext } from 'components/input/Form';
+import * as utils from './utils';
+
+type Props = {
+  /** name prefix of the field to map in the form model.
```
Comment (Etheryte) [inline web/html/src/manager/virtualization/nets/properties/NetworkAddress.js:9]: It's very nice that you comment all of these descriptively, these have helped me a lot with migration stuff in other places and are very nice in general. 👍

File: web/html/src/manager/virtualization/nets/properties/VirtualPortFields.js [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+          labelClass="col-md-3"
+          divClass="col-md-6"
+          defaultValue='profileid'
+          items={[
+            {label: t('By profile id'), value: 'profileid'},
+            {label: t('Virtual Station Interface (VSI) parameters'), value: 'vsi'},
+          ]}
+        />)
+      }
+      { ("openvswitch" === props.type || ("802.1qbh" === props.type && model['virtualport-params-type'] === "profileid")) && (
```
Comment (Etheryte) [inline web/html/src/manager/virtualization/nets/properties/VirtualPortFields.js:34]: Very small nitpick, I don't really have a strong preference on which way it is, please prefer either `props.type === "foo"` or `"foo" === props.type` consistently, right now it's one way in one place and the other way in the other.

File: web/html/src/manager/virtualization/nets/properties/utils.js [hunk 12]
```diff
@@ -0,0 +1,24 @@
+const decByte = '(?:(?:25[0-5])|(?:2[0-4][0-9])|(?:1[0-9]{2})|(?:[1-9][0-9])|(?:[0-9]))';
+
+const ipv4Regex = `(?:${decByte}\\.){3}${decByte}`;
+export const ipv4Pattern = new RegExp(`^${ipv4Regex}$`);
+
+const hexDigit = '[0-9A-Fa-f]';
+const hexWord = `${hexDigit}{1,4}`;
+const ipv6Regex = `(?:(?:${hexWord}:){7}${hexWord})|(?:(?:${hexWord}:){6}:${hexWord})|(?:(?:${hexWord}:){5}:(?:${hexWord}:)?${hexWord})|(?:(?:${hexWord}:){4}:(?:${hexWord}:){0,2}${hexWord})|(?:(?:${hexWord}:){3}:(?:${hexWord}:){0,3}${hexWord})|(?:(?:${hexWord}:){2}:(?:${hexWord}:){0,4}${hexWord})|(?:(?:${hexWord}:){6}(?:${decByte}\\.){3}${decByte})|(?:(?:${hexWord}:){0,5}:(?:${decByte}\\.){3}${decByte})|(?:::(?:${hexWord}:){0,5}(?:${decByte}\\.){3}${decByte})|(?:${hexWord}::(?:${hexWord}:){0,5}${hexWord})|(?:::(?:${hexWord}:){0,6}${hexWord})|(?:(?:${hexWord}:){1,7}:)|(?:::)`;
```
Comment (Etheryte) [inline web/html/src/manager/virtualization/nets/properties/utils.js:8]: Please add test cases for everything exported from this file, there is no way to realistically review the correctness of this and other similar regexes otherwise.

File: web/html/src/manager/virtualization/nets/network-properties.js (hunk 7 above)
Comment (cbosdo) [reply web/html/src/manager/virtualization/nets/network-properties.js:82]: the `unflattenModel()` would think those should be part of an array due to the number at the end of the name. To work this around at no code, I just ensured the name doesn't end with the number before that call and set the names right before sending the request. Not super elegant for sure, but avoids complex changes in the `unflattenModel()` function

File: web/html/src/manager/virtualization/nets/properties/NetworkAddress.js (hunk 10 above)
Comment (cbosdo) [reply web/html/src/manager/virtualization/nets/properties/NetworkAddress.js:9]: yeah, and those are picked up by storybook docs :)

File: web/html/src/manager/virtualization/nets/properties/utils.js (hunk 12 above)
Comment (cbosdo) [reply web/html/src/manager/virtualization/nets/properties/utils.js:8]: Those are picked up from libvirt schema... I would be really surprised there is a mistake in them... and anyway I won't be able to unit test all corner cases of these, but yes, I can add a few ones.

File: web/html/src/components/input/Form.tsx (hunk 2 above)
Comment (cbosdo) [reply web/html/src/components/input/Form.tsx:147]: I'll use `[]` it's fine.

File: branding/css/theme.less (hunk 1 above)
Comment (cbosdo) [reply branding/css/theme.less:905]: OK, I added the import in the global less files though, with the other imports.

File: web/html/src/components/input/FormGroup.tsx (hunk 3 above)
Comment (cbosdo) [reply web/html/src/components/input/FormGroup.tsx:17]: Right, IIRC flow wasn't always happy without those, but in TS it seems OK: makes more sense to me at least.

File: web/html/src/components/input/InputBase.tsx (hunk 4 above)
Comment (cbosdo) [reply web/html/src/components/input/InputBase.tsx:181]: ACK

Comment (cbosdo) [reply web/html/src/components/input/InputBase.tsx:181]: OK

File: web/html/src/components/input/InputBase.tsx (hunk 5 above)
Comment (cbosdo) [reply web/html/src/components/input/InputBase.tsx:182]: OK, I then changed the expression to get rid of the ternary operator:
```ts
_isNil(value) || (Array.isArray(this.props.name) && Object.values(value).filter(v => !this.isEmptyValue(v)).length === 0)
```

File: web/html/src/components/input/form-utils.ts (hunk 6 above)
Comment (cbosdo) [reply web/html/src/components/input/form-utils.ts:81]: OK

Comment (cbosdo) [reply web/html/src/components/input/form-utils.ts:81]: ACK

File: web/html/src/manager/virtualization/nets/network-properties.js (hunk 8 above)
Comment (cbosdo) [reply web/html/src/manager/virtualization/nets/network-properties.js:168]: ACK

File: web/html/src/manager/virtualization/nets/network-properties.js (hunk 9 above)
Comment (cbosdo) [reply web/html/src/manager/virtualization/nets/network-properties.js:277]: ACK

File: web/html/src/manager/virtualization/nets/properties/VirtualPortFields.js (hunk 11 above)
Comment (cbosdo) [reply web/html/src/manager/virtualization/nets/properties/VirtualPortFields.js:34]: ACK

File: web/html/src/manager/virtualization/nets/properties/utils.js (hunk 12 above)
Comment (cbosdo) [reply web/html/src/manager/virtualization/nets/properties/utils.js:8]: added! Thanks... that made me spot a typo in the UUID one ;)

**Comment by cbosdo** [comment]
@Etheryte addressed all your comments

File: web/html/src/components/input/Form.tsx [hunk 13]
```diff
@@ -137,12 +137,14 @@ export class Form extends React.Component<Props> {
             }
             return filtered;
           }, {});
-          const errors = Object.keys(this.props.errors || {}).reduce((filtered, key) => {
-            if (names.includes(key) && this.props.errors?.[key]) {
-              return filtered.concat(this.props.errors[key]);
-            }
-            return filtered;
-          }, []);
+          const errors = this.props.errors != null ?
```
Comment (Etheryte) [inline web/html/src/components/input/Form.tsx:140]: This ternary should be redundant with the latest updates, `Object.keys(this.props.errors || {}).reduce((filtered, key) => filtered, [])` will return `[]` in case that `this.props.errors` is null.

**Comment by Etheryte** [comment]
The frontend parts look good to me, good job on pulling this big piece in. 👍

**Review by Etheryte — APPROVED** [review]

File: web/html/src/components/input/Form.tsx (hunk 13 above)
Comment (Etheryte) [reply web/html/src/components/input/Form.tsx:140]: To perhaps explain what I mean better, the current code:

```ts
const errors = this.props.errors != null ?
    Object.keys(this.props.errors || {}).reduce((filtered, key) => {
        if (names.includes(key) && this.props.errors?.[key]) {
            return filtered.concat(this.props.errors[key]);
        }
        return filtered;
    }, []) :
[];
```

Is equivalent to this code without the ternary:  

```ts
const errors = Object.keys(this.props.errors || {}).reduce((filtered, key) => {
    if (names.includes(key) && this.props.errors?.[key]) {
        return filtered.concat(this.props.errors[key]);
    }
    return filtered;
}, []);
```

Comment (cbosdo) [reply web/html/src/components/input/Form.tsx]: OK, just fixed it

File: web/html/src/components/input/Text.tsx [hunk 14]
```diff
@@ -34,6 +34,7 @@ export const Text = (props: Props) => {
             className={`form-control${inputClass ? ` ${inputClass}` : ""}`}
             type={type || "text"}
             name={props.name}
+            id={props.name}
```
Comment (Etheryte) [inline web/html/src/components/input/Text.tsx:37]: This isn't really safe and may lead to unexpected bugs. The problem here is that the name property could be anything and it isn't clear from the outside that it will also be used as an id. To exemplify, it would be perfectly valid to have a field with a name of say `network` but also have a DOM node with the same id somewhere else on the page, perhaps in a completely unrelated component. This would then cause all sorts of unexpected problems.  

The fix is simple, `id` should be a prop by itself and it should be explicitly set when needed.

Comment (cbosdo) [reply web/html/src/components/input/Text.tsx:37]: The `for` part is generated [here](https://github.com/uyuni-project/uyuni/blob/master/web/html/src/components/input/InputBase.tsx#L255) in the `InputBase` component and uses the field name as well. This is already done this way in other form components, just that this one was left aside for some reason.

If someone wants to fix this at the `InputBase` level and in all the forms using it why not... but I won't start such a task myself.

Comment (Etheryte) [reply web/html/src/components/input/Text.tsx:37]: Oh okay, that makes me sad. In that case this is okay, I'll create a ticket to clean this up in the future.

Comment (Etheryte) [reply web/html/src/components/input/Text.tsx:37]: Created https://github.com/SUSE/spacewalk/issues/14165 to track this.

Comment (cbosdo) [reply web/html/src/components/input/Text.tsx:37]: thanks for tracking this!

File: web/html/src/utils/test-utils/index.js [hunk 15]
```diff
@@ -35,6 +40,20 @@ export const type = async (elementOrPromiseOfElement, text, options) => {
   return new Promise(resolve => window.requestAnimationFrame(() => resolve()));
 }
 
+/** Get a form field object */
+export const getFieldByName = (formName, fieldName) => {
+  return [...screen.getByTitle(formName).elements]
```
Comment (Etheryte) [inline web/html/src/utils/test-utils/index.js:45]: Does the spread (`[...foo]`) solve a problem here over say `screen.getByTitle(formName).elements.filter(...)`? Right now it isn't very clear, unless I'm missing something here?

File: web/html/src/manager/virtualization/nets/properties/fields-data.js [hunk 16]
```diff
@@ -0,0 +1,55 @@
+import * as data from 'utils/data';
+
+export const mapping = {
+    bridge: {
+      description: 'Use an existing and configured bridge device. ' +
```
Comment (Etheryte) [inline web/html/src/manager/virtualization/nets/properties/fields-data.js:5]: I'm guessing all of these should probably be wrapped in `t()`?

File: web/html/src/manager/virtualization/nets/properties/Vlans.js [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+            formContext.setModelValue(`vlans${index}_tag`, '');
+            if (index > 0) {
+              formContext.setModelValue(`vlantrunk`, true);
+            }
+          }}
+          onRemove={(index: number) => {
+            Object.keys(model)
+              .filter(key => key.startsWith(`vlans${index}_`))
+              .forEach(key => formContext.setModelValue(key, undefined));
+            if (Object.keys(model).filter(key => key.match(/^vlan[0-9]+_tag$/)).length <= 1) {
```
Comment (Etheryte) [inline web/html/src/manager/virtualization/nets/properties/Vlans.js:32]: This line could use a comment describing what the logic here is, without any context it's a bit confusing why we want to do this on both 0 and 1 matches for example.

File: web/html/src/components/input/form-utils.ts [hunk 18]
```diff
@@ -63,7 +91,8 @@ export function unflattenModel<T>(flatModel: Record<string, T>): TreeLikeModel<T
   const mergeArrays = (model: any): any => {
     return Object.keys(model).reduce((result, key) => {
       const matcher = key.match(/^([a-zA-Z0-9]*[A-zA-Z])[0-9]+$/);
-      const mergedValue = model[key] instanceof Object ? mergeArrays(model[key]) : model[key];
+      const mergedValue = typeof model[key] === "object" && !Array.isArray(model[key]) ?
```
Comment (Etheryte) [inline web/html/src/components/input/form-utils.ts:94]: I'm guessing there was a bug here with handling arrays since they're also objects? It would be nice to add one small test case that covers the case that was bugged before but works now. Nice catch. 👍

File: web/html/src/manager/virtualization/nets/network-properties.test.js [hunk 19]
```diff
@@ -0,0 +1,367 @@
...
+            {address: "192.168.1.1", domain: "example.com"},
+            {domain: "acme.com"},
+          ],
+        },
+      });
+    };
+
+    await renderWithNetwork();
+    await select(screen.getByLabelText(/^Network type/), "open")
+    fireEvent.change(screen.getByLabelText(/^Name/), {target: {value: "open0"}});
```
Comment (Etheryte) [inline web/html/src/manager/virtualization/nets/network-properties.test.js:181]: If this works and the current type implementation doesn't, I would move this logic into `type()` instead and verify that it works everywhere. Since it isn't clear without context why this logic is used here but `type()` elsewhere, it's unlikely someone else would come around and fix this later.

**Comment by Etheryte** [comment]
I've reviewed the frontend parts again, with just a few small notes. Other than those I think it would be good to get this merged soon, even if at a halfway point. The PR is pretty big and it would be easier to review changes in smaller portions.

File: web/html/src/manager/virtualization/nets/properties/fields-data.js (hunk 16 above)
Comment (cbosdo) [reply web/html/src/manager/virtualization/nets/properties/fields-data.js:5]: The `t()` call is where that value is used: 
```
hint={t(FieldsData.getValue(model.type, 'description', ''))}
```
But may be it would be better for automatic strings gathering to move it here indeed.

File: web/html/src/components/input/form-utils.ts (hunk 18 above)
Comment (cbosdo) [reply web/html/src/components/input/form-utils.ts:94]: Well, I modified a test for that. See here: https://github.com/uyuni-project/uyuni/pull/3286/files#diff-5eb60c337c8f8aad296dac8a0a9a8d7ee1ec89c89ae9c51b5b64623af90a02dfL74-R87

File: web/html/src/manager/virtualization/nets/properties/Vlans.js (hunk 17 above)
Comment (cbosdo) [reply web/html/src/manager/virtualization/nets/properties/Vlans.js:32]: ACK

File: web/html/src/utils/test-utils/index.js [hunk 20]
```diff
@@ -35,6 +40,20 @@ export const type = async (elementOrPromiseOfElement, text, options) => {
   return new Promise(resolve => window.requestAnimationFrame(() => resolve()));
 }
 
+/** Get a form field object */
+export const getFieldByName = (formName, fieldName) => {
```
[the remark this answers is not in the document]
Comment (cbosdo) [reply web/html/src/utils/test-utils/index.js:44]: how would you see those? reexported by the index? Note that the `datetimepicker` mock needs to be run by the index before the test file is actually running through its imports

File: web/html/src/utils/test-utils/index.js (hunk 15 above)
Comment (cbosdo) [reply web/html/src/utils/test-utils/index.js:45]: This is because `screen.getByTitle(formName).elements` is not an Array, but an `HTMLFormControlsCollection`. Spreading it converts it into an Array (found the trick on the web).

File: web/html/src/manager/virtualization/nets/network-properties.test.js (hunk 19 above)
Comment (cbosdo) [reply web/html/src/manager/virtualization/nets/network-properties.test.js:181]: on that one I have mixed feelings... indeed it would be better to use `type()` and fix it. It would also avoid depending on the components implementation in the tests, but on the other hand tests miserably fail when using `type()`...

File: web/html/src/components/input/form-utils.ts (hunk 18 above)
Comment (Etheryte) [reply web/html/src/components/input/form-utils.ts:94]: Ah okay, I missed that, sorry. Good fix.

File: web/html/src/utils/test-utils/index.js (hunk 15 above)
Comment (Etheryte) [reply web/html/src/utils/test-utils/index.js:45]: Sounds good, makes sense. 👍

File: web/html/src/utils/test-utils/index.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (Etheryte) [reply web/html/src/utils/test-utils/index.js:44]: Reexporting works well yeah, I pushed a small demo to this branch. Feel free to roll back if this does not look good for you.

**Review by Etheryte — CHANGES_REQUESTED** [review]
I had a look at the test failures, the main problem seems to be that the UI hasn't updated yet when checking for fields. The loading issue is one of those, another one is fields appearing when you trigger click events etc. I would recommend fixing these before merging this because otherwise this test will be very fragile and can break at some unrelated change.  

In broad strokes, the procedure is to replace all `fireEvent.change` back to `type` (after you've rebased the updates to `type` from master), then see where you run into timing issues. If you have loading screens, you can use `waitForElementToBeRemoved` to wait for loading to be done (if there's no better way to check in your async flow). For other elements, you might want to consider different queries depending on the situation (see https://testing-library.com/docs/queries/about/#types-of-queries). If you use a query that synchronously checks whether an element is there and it isn't and error will be thrown.  

Let me know if the above is unclear or if you need help with some aspect of it.

**Review by Etheryte — APPROVED** [review]
Now that the test issues are resolved, this looks good to go. 👍

**Review by mbologna — APPROVED** [review]
